Fix docker cp when container source path is /

tiborvass opened the pull request:
Another attempt at fixing #39348
Fixes #39348
Previous attempt at #39351

Before 7a7357d, archive.TarResourceRebase was being used to copy files
and folders from the container. That function splits the source path
into a dirname + basename pair to support copying a file:
if you wanted to tar `dir/file` it would tar from `dir` the file `file`
(as part of the IncludedFiles option).

However, that path splitting logic was kept for folders as well, which
resulted in weird inputs to archive.TarWithOptions:
if you wanted to tar `dir1/dir2` it would tar from `dir1` the directory
`dir2` (as part of IncludedFiles option).

Although it was weird, it worked fine until we started chrooting into
the container rootfs when doing a `docker cp` with container source set
to `/` (cf 3029e76 (https://github.com/moby/moby/pull/39292)).

The fix is to only do the path splitting logic if the source is a file.

Unfortunately, 7a7357d added support for LCOW by duplicating some of
this subtle logic. Ideally we would need to do more refactoring of the
archive codebase to properly encapsulate these behaviors behind well-
documented APIs.

This fix does not do that. Instead, it fixes the issue inline.

Signed-off-by: Tibor Vass <tibor@docker.com>


I added a couple of more tests than the actual issue needs, just to make sure there are no other regressions compared to before the cve fix (3029e76).

Huge thanks to @cpuguy83 ❤️who worked tirelessly with me to understand the code and make this PR.

Commit:
daemon: fix docker cp when container source is /

Before 7a7357da, archive.TarResourceRebase was being used to copy files
and folders from the container. That function splits the source path
into a dirname + basename pair to support copying a file:
if you wanted to tar `dir/file` it would tar from `dir` the file `file`
(as part of the IncludedFiles option).

However, that path splitting logic was kept for folders as well, which
resulted in weird inputs to archive.TarWithOptions:
if you wanted to tar `dir1/dir2` it would tar from `dir1` the directory
`dir2` (as part of IncludedFiles option).

Although it was weird, it worked fine until we started chrooting into
the container rootfs when doing a `docker cp` with container source set
to `/` (cf 3029e765).

The fix is to only do the path splitting logic if the source is a file.

Unfortunately, 7a7357da added support for LCOW by duplicating some of
this subtle logic. Ideally we would need to do more refactoring of the
archive codebase to properly encapsulate these behaviors behind well-
documented APIs.

This fix does not do that. Instead, it fixes the issue inline.

Signed-off-by: Tibor Vass <tibor@docker.com>

cpuguy83 committed:
Add test for copying entire container rootfs

CID=$(docker create alpine)
docker cp $CID:/ out

Signed-off-by: Brian Goff <cpuguy83@gmail.com>
Signed-off-by: Tibor Vass <tibor@docker.com>

cpuguy83 commented:
```
15:29:30 --- FAIL: TestCopyFromContainerRoot (1.83s)
15:29:30     copy_test.go:146: assertion failed: expression is false: found["/foo"]: /foo file not found in archive
15:29:30     copy_test.go:147: assertion failed: expression is false: found["/bar/baz"]: /bar/baz file not found in archive
```

Commit:
pkg/archive: keep walkRoot clean if source is /

Previously, getWalkRoot("/", "foo") would return "//foo"
Now it returns "/foo"

Signed-off-by: Tibor Vass <tibor@docker.com>

Commit:
add more tests

Signed-off-by: Tibor Vass <tibor@docker.com>

tiborvass commented:
Ping @cpuguy83 @thaJeztah @tonistiigi @justincormack

tiborvass commented on pkg/archive/archive_unix.go:
```diff
@@ -26,6 +26,9 @@ func fixVolumePathPrefix(srcPath string) string {
 // can't use filepath.Join(srcPath,include) because this will clean away
 // a trailing "." or "/" which may be important.
 func getWalkRoot(srcPath string, include string) string {
+	if srcPath == string(filepath.Separator) {
```
FWIW this is not needed for the main fix. It is just fixing a weird input (`getWalkRoot("/", ".")` used to return `//.`, now just `/.`)

kolyshkin replied:
Hmm, why can't we just use `path.Join(srcPath, include)` which should handle this and other similar cases? Say, https://play.golang.org/p/jG-JpZ-Kun5

kolyshkin replied:
Nevermind, I missed the comment explaining why :man_facepalming:

thaJeztah commented on daemon/archive.go:
```diff
@@ -426,9 +432,6 @@ func (daemon *Daemon) containerCopy(container *container.Container, resource str
 		basePath = d
 		filter = []string{f}
-	} else {
-		filter = []string{driver.Base(basePath)}
```
No filter is needed in this case?

tiborvass replied:
This was to keep it symmetrical and I believe is used by older docker CLIs (this codepath is dead code for newer APIs, but not older ones). @cpuguy83 confirmed as well that cp works as intended for an older docker CLI (like 1.8).

tiborvass replied:
Correct, because no filter means everything.

cpuguy83 reviewed (approved):
LGTM

thaJeztah commented on pkg/archive/archive_unix.go:
```diff
@@ -26,6 +26,9 @@ func fixVolumePathPrefix(srcPath string) string {
+	if srcPath == string(filepath.Separator) {
+		return srcPath + include
+	}
 	return srcPath + string(filepath.Separator) + include
```
^^ not a blocker, but worth considering

tiborvass replied:
In that case I would use strings.Index to determine if there is one, instead of removing it to add it back.

thaJeztah replied:
Could work as well; thought there would not be much overhead in normalizing and unconditionally adding

tonistiigi commented on daemon/archive.go:
```diff
@@ -236,7 +236,13 @@ func (daemon *Daemon) containerArchivePath(container *container.Container, path
+	sourceDir := resolvedPath
+	sourceBase := "."
+
+	if stat.Mode&os.ModeDir == 0 { // not dir
```
I think this should be

```
if symlink {
opts = rebase
}
```

and always `archivePath(resolvedPath)`.

Would be more understandable what is actually happening and wouldn't need the hack for removing the slash in the walker. But functionally they seem equivalent.

tiborvass replied:
Agreed. Can be done in a follow up.

tiborvass commented:
@thaJeztah @kolyshkin updated with small requested change.

tiborvass commented:
Failure is TestAPISwarmLeaderElection which is unrelated

thaJeztah reviewed (approved):
LGTM

thaJeztah commented:
Looks like s390x passed, but is hanging on the cleanup; https://jenkins.dockerproject.org/job/Docker-PRs-s390x/14591/console

```
06:06:03 OK: 1421 passed, 81 skipped
06:06:03 PASS
06:06:03 ---> Making bundle: .integration-daemon-stop (in bundles/test-integration)
06:06:16 Clearing AppArmor profiles cache:.
06:06:16 All profile caches have been cleared, but no profiles have been unloaded.
06:06:16 Unloading profiles will leave already running processes permanently
06:06:16 unconfined, which can lead to unexpected situations.
06:06:16 
06:06:16 To set a process to complain mode, use the command line tool
06:06:16 'aa-complain'. To really tear down all profiles, run the init script
06:06:16 with the 'teardown' option."
06:06:16 Removing test suite binaries
```
